library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java

```diff
@@ -0,0 +1,490 @@
+  private List<Window.TimeWindow> getTimeWindowsForTimestamp(long timestamp)
+  {
+    List<Window.TimeWindow> windows = new ArrayList<>();
+    if (windowOption instanceof WindowOption.TimeWindows) {
+      long durationMillis = ((WindowOption.TimeWindows)windowOption).getDuration().getMillis();
+      long beginTimestamp = timestamp - timestamp % durationMillis;
+      windows.add(new Window.TimeWindow(beginTimestamp, durationMillis));
+      if (windowOption instanceof WindowOption.SlidingTimeWindows) {
+        long slideBy = ((WindowOption.SlidingTimeWindows)windowOption).getSlideByDuration().getMillis();
+        // add the sliding windows front and back
+        // Note: this messes up the order of the window and we might want to revisit this if the order of the windows
+        // matter
+        for (long slideBeginTimestamp = beginTimestamp - slideBy;
+            slideBeginTimestamp <= timestamp && timestamp < slideBeginTimestamp + durationMillis;
+            slideBeginTimestamp -= slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+        for (long slideBeginTimestamp = beginTimestamp + slideBy;
+            slideBeginTimestamp <= timestamp && timestamp < slideBeginTimestamp + durationMillis;
+            slideBeginTimestamp += slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+      }
+    } else {
+      throw new IllegalStateException("Unexpected WindowOption");
+    }
+    return windows;
+  }
+
+  @Override
+  public boolean isTooLate(long timestamp)
+  {
+    return allowedLatenessMillis < 0 ? false : (timestamp < currentWatermark - allowedLatenessMillis);
+  }
+
+  @Override
+  public void dropTuple(Tuple<InputT> input)
+  {
+    // do nothing
+    LOG.debug("Dropping late tuple {}", input);
+  }
+
+
+  @Override
+  public void processWatermark(ControlTuple.Watermark watermark)
+  {
+    currentWatermark = watermark.getTimestamp();
+    long horizon = currentWatermark - allowedLatenessMillis;
+    if (allowedLatenessMillis >= 0) {
+      // purge window that are too late to accept any more input
+      dataStorage.removeUpTo(horizon);
+    }
+
+    for (Iterator<Map.Entry<Window, WindowState>> it = windowStateMap.iterator(); it.hasNext(); ) {
+      Map.Entry<Window, WindowState> entry = it.next();
+      Window window = entry.getKey();
+      WindowState windowState = entry.getValue();
+      if (allowedLatenessMillis >= 0 && window.getBeginTimestamp() + window.getDurationMillis() < horizon) {
+        // discard this window because it's too late now
+        it.remove();
+      } else if (window.getBeginTimestamp() + window.getDurationMillis() < currentWatermark) {
+        // watermark has not arrived for this window before, marking this window late
+        if (windowState.watermarkArrivalTime == -1) {
+          windowState.watermarkArrivalTime = currentDerivedTimestamp;
+          if (triggerAtWatermark) {
```

Comment:
Shouldn't triggerAtWatermark be a default option?
As per beam specification, trigger at watermark always happens.